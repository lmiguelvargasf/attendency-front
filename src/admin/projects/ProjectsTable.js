import React, { useReducer } from 'react'
import { Link } from 'react-router-dom'
import { Table, Space, Button, Select, message } from 'antd'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faEdit, faCalendar, faUserPlus, faPlus } from '@fortawesome/free-solid-svg-icons'
import axios from 'axios'
import RemoveObjectButton from '../RemoveObjectButton'
import AddMemberModal from './AddMemberModal'

import styles from '../Admin.module.sass'

const addMemberStateReducer = (state, action) => {
  switch (action.type) {
    case 'OPEN_MODAL':
      return { ...state, visible: true, projectKey: action.projectKey, selectedProjectIndex: action.selectedProjectIndex }
    case 'LOADING':
      return { ...state, confirmLoading: true }
    case 'SET_NON_MEMBERS':
      return { ...state, nonMembers: action.nonMembers }
    case 'SET_MEMBER_TO_ADD':
      return { ...state, memberToAdd: action.memberToAdd }
    case 'UNSET_MEMBER_TO_ADD':
      return { ...state, memberToAdd: null }
    case 'CLOSE_MODAL':
      return { ...state, memberToAdd: null, confirmLoading: false, visible: false }
    default:
      throw new Error(`Unhandled action type: ${action.type}`)
  }
}

const ProjectsTable = ({ projects, removeProject, updateProjects }) => {
  const [addMemberState, dispatch] = useReducer(addMemberStateReducer, {
    visible: false,
    confirmLoading: false,
    nonMembers: [],
    projectKey: null,
    selectedProjectIndex: null,
    memberToAdd: null
  })

  const showModal = async (project, index) => {
    dispatch({ type: 'OPEN_MODAL', selectedProjectIndex: index, projectKey: project.key })
    let response
    try {
      response = await axios.get(`${process.env.REACT_APP_API_URL}/projects/${project.key}/non-members/`)
    } catch (error) {
      console.log(error)
    }
    dispatch({ type: 'SET_NON_MEMBERS', nonMembers: response.data })
  }
  const handleOk = async () => {
    if (addMemberState.nonMembers.length === 0) {
      dispatch({ type: 'CLOSE_MODAL' })
      return
    }

    if (!addMemberState.memberToAdd) {
      message.error('Please select a member to add!')
      return
    }
    dispatch({ type: 'LOADING' })
    let response
    try {
      response = await axios.post(`${process.env.REACT_APP_API_URL}/projects/${addMemberState.projectKey}/add-member/`, { key: addMemberState.memberToAdd })
    } catch (error) {
      message.error('There was an error, please try again.')
      console.log(error)
    }
    const updatedProject = response.data
    updateProjects(updatedProject, addMemberState.selectedProjectIndex)
    dispatch({ type: 'CLOSE_MODAL' })
    message.success('Member was added successfully!')
  }

  const handleCancel = () => {
    dispatch({ type: 'CLOSE_MODAL' })
  }
  const handleOnChange = value => {
    dispatch({ type: 'SET_MEMBER_TO_ADD', memberToAdd: value })
  }

  const columns = [
    {
      title: 'Title',
      dataIndex: 'title',
      key: 'title'
    },
    {
      title: 'Start Date',
      dataIndex: 'startDate',
      key: 'startDate'
    },
    {
      title: 'Description',
      dataIndex: 'description',
      key: 'description',
      render: description => {
        if (description) {
          return description.length < 50 ? description : `${description.substring(0, 50)}...`
        }
        return '-'
      }
    },
    {
      title: 'Team',
      dataIndex: 'team',
      key: 'team'
    },
    {
      title: 'Action',
      key: 'action',
      render: (text, record, index) => (
        <Space size='middle'>
          <Link to={{ pathname: `/admin/projects/${record.key}/edit`, state: { project: record, index } }}>
            <FontAwesomeIcon icon={faEdit} />
          </Link>
          <FontAwesomeIcon icon={faCalendar} />
          <button type='button' className={styles.linkButton} onClick={() => showModal(record, index)}>
            <FontAwesomeIcon icon={faUserPlus} style={{ color: '#9664c8' }} />
          </button>
          <RemoveObjectButton object={record} removeObject={removeProject} />
        </Space>
      )
    }
  ]

  return (
    <>
      <Space direction='vertical' size='middle' style={{ width: '100%' }}>
        <Link to='/admin/projects/create'>
          <Button type='primary'>
            <Space>
              <FontAwesomeIcon icon={faPlus} />
              <span>New project</span>
            </Space>
          </Button>
        </Link>
        <Table
          data-testid='project-table'
          columns={columns}
          dataSource={projects}
          pagination={false}
        />
      </Space>
      <AddMemberModal state={addMemberState} handleOk={handleOk} handleCancel={handleCancel} handleOnChange={handleOnChange} />
    </>
  )
}

export default ProjectsTable
