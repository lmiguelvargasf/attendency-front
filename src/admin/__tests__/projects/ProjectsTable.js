import React from 'react'
import { BrowserRouter as Router } from 'react-router-dom'
import { render, cleanup } from '@testing-library/react'
import ProjectsTable from '../../projects/ProjectsTable'

Object.defineProperty(window, 'matchMedia', {
  writable: true,
  value: jest.fn().mockImplementation(query => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // deprecated
    removeListener: jest.fn(), // deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn()
  }))
})

describe('ProjectsTable component', () => {
  afterEach(cleanup)

  it('matches snapshot no projects', () => {
    const { asFragment } = render(
      <Router><ProjectsTable projects={[]} removeProject={() => {}} /></Router>
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it('matches snapshot one project', () => {
    const BASE_API_URL = process.env.REACT_APP_API_URL
    const project = {
      key: 1,
      url: `${BASE_API_URL}/projects/1`,
      title: 'Testing Project Alpha',
      startDate: '2020-04-18',
      description: 'This is just for testing',
      team: 'A, B, C'
    }
    const { asFragment } = render(
      <Router><ProjectsTable projects={[project]} removeProject={() => {}} /></Router>
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
