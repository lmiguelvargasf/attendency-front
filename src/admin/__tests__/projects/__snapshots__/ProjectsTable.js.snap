// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectsTable component matches snapshot no projects 1`] = `
<DocumentFragment>
  <div
    class="ant-space ant-space-vertical"
    style="width: 100%;"
  >
    <div
      class="ant-space-item"
      style="margin-bottom: 16px;"
    >
      <a
        href="/admin/projects/create"
      >
        <button
          class="ant-btn ant-btn-primary"
          type="button"
        >
          <div
            class="ant-space ant-space-horizontal"
          >
            <div
              class="ant-space-item"
              style="margin-right: 8px;"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-plus fa-w-14 "
                data-icon="plus"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 448 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
                  fill="currentColor"
                />
              </svg>
            </div>
            <div
              class="ant-space-item"
            >
              <span>
                New project
              </span>
            </div>
          </div>
        </button>
      </a>
    </div>
    <div
      class="ant-space-item"
    >
      <div
        class="ant-table-wrapper"
      >
        <div
          class="ant-spin-nested-loading"
        >
          <div
            class="ant-spin-container"
          >
            <div
              class="ant-table"
              data-testid="project-table"
            >
              <div
                class="ant-table-container"
              >
                <div
                  class="ant-table-content"
                >
                  <table
                    style="table-layout: auto;"
                  >
                    <colgroup />
                    <thead
                      class="ant-table-thead"
                    >
                      <tr>
                        <th
                          class="ant-table-cell"
                        >
                          Title
                        </th>
                        <th
                          class="ant-table-cell"
                        >
                          Start Date
                        </th>
                        <th
                          class="ant-table-cell"
                        >
                          Description
                        </th>
                        <th
                          class="ant-table-cell"
                        >
                          Team
                        </th>
                        <th
                          class="ant-table-cell"
                        >
                          Action
                        </th>
                      </tr>
                    </thead>
                    <tbody
                      class="ant-table-tbody"
                    >
                      <tr
                        class="ant-table-placeholder"
                      >
                        <td
                          class="ant-table-cell"
                          colspan="5"
                        >
                          <div
                            class="ant-empty ant-empty-normal"
                          >
                            <div
                              class="ant-empty-image"
                            >
                              <svg
                                class="ant-empty-img-simple"
                                height="41"
                                viewBox="0 0 64 41"
                                width="64"
                                xmlns="http://www.w3.org/2000/svg"
                              >
                                <g
                                  fill="none"
                                  fill-rule="evenodd"
                                  transform="translate(0 1)"
                                >
                                  <ellipse
                                    class="ant-empty-img-simple-ellipse"
                                    cx="32"
                                    cy="33"
                                    rx="32"
                                    ry="7"
                                  />
                                  <g
                                    class="ant-empty-img-simple-g"
                                    fill-rule="nonzero"
                                  >
                                    <path
                                      d="M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"
                                    />
                                    <path
                                      class="ant-empty-img-simple-path"
                                      d="M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z"
                                    />
                                  </g>
                                </g>
                              </svg>
                            </div>
                            <p
                              class="ant-empty-description"
                            >
                              No Data
                            </p>
                          </div>
                        </td>
                      </tr>
                    </tbody>
                  </table>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ProjectsTable component matches snapshot one project 1`] = `
<DocumentFragment>
  <div
    class="ant-space ant-space-vertical"
    style="width: 100%;"
  >
    <div
      class="ant-space-item"
      style="margin-bottom: 16px;"
    >
      <a
        href="/admin/projects/create"
      >
        <button
          class="ant-btn ant-btn-primary"
          type="button"
        >
          <div
            class="ant-space ant-space-horizontal"
          >
            <div
              class="ant-space-item"
              style="margin-right: 8px;"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-plus fa-w-14 "
                data-icon="plus"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 448 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
                  fill="currentColor"
                />
              </svg>
            </div>
            <div
              class="ant-space-item"
            >
              <span>
                New project
              </span>
            </div>
          </div>
        </button>
      </a>
    </div>
    <div
      class="ant-space-item"
    >
      <div
        class="ant-table-wrapper"
      >
        <div
          class="ant-spin-nested-loading"
        >
          <div
            class="ant-spin-container"
          >
            <div
              class="ant-table"
              data-testid="project-table"
            >
              <div
                class="ant-table-container"
              >
                <div
                  class="ant-table-content"
                >
                  <table
                    style="table-layout: auto;"
                  >
                    <colgroup />
                    <thead
                      class="ant-table-thead"
                    >
                      <tr>
                        <th
                          class="ant-table-cell"
                        >
                          Title
                        </th>
                        <th
                          class="ant-table-cell"
                        >
                          Start Date
                        </th>
                        <th
                          class="ant-table-cell"
                        >
                          Description
                        </th>
                        <th
                          class="ant-table-cell"
                        >
                          Team
                        </th>
                        <th
                          class="ant-table-cell"
                        >
                          Action
                        </th>
                      </tr>
                    </thead>
                    <tbody
                      class="ant-table-tbody"
                    >
                      <tr
                        class="ant-table-row ant-table-row-level-0"
                        data-row-key="1"
                      >
                        <td
                          class="ant-table-cell"
                        >
                          Testing Project Alpha
                        </td>
                        <td
                          class="ant-table-cell"
                        >
                          2020-04-18
                        </td>
                        <td
                          class="ant-table-cell"
                        >
                          This is just for testing
                        </td>
                        <td
                          class="ant-table-cell"
                        >
                          A, B, C
                        </td>
                        <td
                          class="ant-table-cell"
                        >
                          <div
                            class="ant-space ant-space-horizontal"
                          >
                            <div
                              class="ant-space-item"
                              style="margin-right: 16px;"
                            >
                              <a
                                href="/admin/projects/1/edit"
                              >
                                <svg
                                  aria-hidden="true"
                                  class="svg-inline--fa fa-edit fa-w-18 "
                                  data-icon="edit"
                                  data-prefix="fas"
                                  focusable="false"
                                  role="img"
                                  viewBox="0 0 576 512"
                                  xmlns="http://www.w3.org/2000/svg"
                                >
                                  <path
                                    d="M402.6 83.2l90.2 90.2c3.8 3.8 3.8 10 0 13.8L274.4 405.6l-92.8 10.3c-12.4 1.4-22.9-9.1-21.5-21.5l10.3-92.8L388.8 83.2c3.8-3.8 10-3.8 13.8 0zm162-22.9l-48.8-48.8c-15.2-15.2-39.9-15.2-55.2 0l-35.4 35.4c-3.8 3.8-3.8 10 0 13.8l90.2 90.2c3.8 3.8 10 3.8 13.8 0l35.4-35.4c15.2-15.3 15.2-40 0-55.2zM384 346.2V448H64V128h229.8c3.2 0 6.2-1.3 8.5-3.5l40-40c7.6-7.6 2.2-20.5-8.5-20.5H48C21.5 64 0 85.5 0 112v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V306.2c0-10.7-12.9-16-20.5-8.5l-40 40c-2.2 2.3-3.5 5.3-3.5 8.5z"
                                    fill="currentColor"
                                  />
                                </svg>
                              </a>
                            </div>
                            <div
                              class="ant-space-item"
                              style="margin-right: 16px;"
                            >
                              <svg
                                aria-hidden="true"
                                class="svg-inline--fa fa-calendar fa-w-14 "
                                data-icon="calendar"
                                data-prefix="fas"
                                focusable="false"
                                role="img"
                                viewBox="0 0 448 512"
                                xmlns="http://www.w3.org/2000/svg"
                              >
                                <path
                                  d="M12 192h424c6.6 0 12 5.4 12 12v260c0 26.5-21.5 48-48 48H48c-26.5 0-48-21.5-48-48V204c0-6.6 5.4-12 12-12zm436-44v-36c0-26.5-21.5-48-48-48h-48V12c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v52H160V12c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v52H48C21.5 64 0 85.5 0 112v36c0 6.6 5.4 12 12 12h424c6.6 0 12-5.4 12-12z"
                                  fill="currentColor"
                                />
                              </svg>
                            </div>
                            <div
                              class="ant-space-item"
                              style="margin-right: 16px;"
                            >
                              <button
                                class="linkButton"
                                type="button"
                              >
                                <svg
                                  aria-hidden="true"
                                  class="svg-inline--fa fa-user-plus fa-w-20 "
                                  data-icon="user-plus"
                                  data-prefix="fas"
                                  focusable="false"
                                  role="img"
                                  style="color: rgb(150, 100, 200);"
                                  viewBox="0 0 640 512"
                                  xmlns="http://www.w3.org/2000/svg"
                                >
                                  <path
                                    d="M624 208h-64v-64c0-8.8-7.2-16-16-16h-32c-8.8 0-16 7.2-16 16v64h-64c-8.8 0-16 7.2-16 16v32c0 8.8 7.2 16 16 16h64v64c0 8.8 7.2 16 16 16h32c8.8 0 16-7.2 16-16v-64h64c8.8 0 16-7.2 16-16v-32c0-8.8-7.2-16-16-16zm-400 48c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z"
                                    fill="currentColor"
                                  />
                                </svg>
                              </button>
                            </div>
                            <div
                              class="ant-space-item"
                            >
                              <a
                                data-testid="http://localhost:8000/api/projects/1"
                              >
                                <svg
                                  aria-hidden="true"
                                  class="svg-inline--fa fa-trash fa-w-14 "
                                  data-icon="trash"
                                  data-prefix="fas"
                                  focusable="false"
                                  role="img"
                                  style="color: rgb(242, 70, 61);"
                                  viewBox="0 0 448 512"
                                  xmlns="http://www.w3.org/2000/svg"
                                >
                                  <path
                                    d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
                                    fill="currentColor"
                                  />
                                </svg>
                              </a>
                            </div>
                          </div>
                        </td>
                      </tr>
                    </tbody>
                  </table>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
